### ISSUE TITLE
hyper parameter research

### ISSUE THREAD

**Issue by bettyeats** [bettyeats/AKBNet#3]
is there any more effecitve method than starting form a learing rate of 0.1? if there is, why?
besides learing rate, how to weight_decay, momentum, etc.

**Comment by bettyeats** [comment]
hyper parameters below:
learning rate;
weight_decay;
momentum;
base_lr;
lr_policy;
gamma;
power;

**Comment by bettyeats** [comment]
lr_policy can be setted as below:

- fixed
keep base_lr stable
学习率不变
- step
if lr_policy is step, stepsize must be setted. Return base_lr\*gamma^(floor(iter/stepsize)). iter represents iterations.
设置学习策略为step，那么还要设置一个参数stepsize。返回 base_lr\*gamma^(floor(iter/stepsize))。iter表示迭代次数。
- exp
Return base_lr\*gamma^iter. iter means iterations.
返回 base_lr\*gamma^iter。iter表示迭代次数。
- inv
if lr_policy is inv, power must be setted. Return base_lr\*(1+gamma*iter)^(-power)
如果学习策略是inv，还要设置参数power。返回base_lr\*(1+gamma*iter)^(-power)。
- multistep
if lr_policy is multistep, stepvalue must be setted.  Changed by stepvalue(more than one).
如果学习策略是multistep，还要设置参数stepvalue。和step一样，随着stepvalue改变，不同的是step策略，stepsize是固定的。
- poly
Return base_lr(1-iter/max_iter)^(power).
返回base_lr(1-iter/max_iter)^(power)。
- sigmoid
Learning rate attenuates in the way of sigmoid. Return base_lr(1/(1+exp(-gamma\*(iter-stepsize))))
用sigmoid方式衰减。返回base_lr(1/(1+exp(-gamma\*(iter-stepsize))))。

**Comment by bettyeats** [comment]
- base_lr
The stepsize of gradient descenting. Too small caused local optimum as well as too big caused passing the lowest point.
-  weight_decay
To prevent over-fitting.
- max_iter
Max iterations. Too small caused misconvergence as well as too big caused oscillation.
- snapshot
Save the snapshot. If set snapshot_diff, save the gradient.
- solver_mode
Set mode. Operate on CPU or GPU.

**Comment by bettyeats** [comment]
- test_iter
Depended on batch_size. A epoch means all data has been implemented.
- test_interval
Interval between test and train.

**Comment by bettyeats** [comment]
调参的技巧：

- 训练集，测试集准确率都很低，那么有可能欠拟合。训练集高，测试集低，模型有可能过拟合。欠拟合，考虑增强模型的拟合能力，如，增加网络层数、增加网络节点数，减少dropout值，减少L2正则值。过拟合，则提高模型泛化能力，来调节参数。

- 在小数据集上效果好的参数，在大数据集上效果也不会太差。可以简化数据，提高速度。比如，对训练数据进行采样，或减少训练类别。

**Comment by bettyeats** [comment]
一些参数的经验值：

- 学习率：1 0.1 0.01 0.0001，大于10的很少见

- 衰减系数：0.5，衰减时机，可以是验证集准确率不再上升时，或固定训练多少个周期以后。

- 网络层数：先从1层开始

- 每层节点数：16 32 128，超过1000的比较少见，超过1w的从来没见过

- batch size：128上下开始。batch size较大，一般不会对结果有太大影响，可以考虑从一个比较大的值开始尝试。batch size太小，有可能结果很差。

**Comment by bettyeats** [comment]
总结：

- 合理性检查，确定模型，数据和其他地方没有问题。

- 训练时跟踪损失函数值，训练集和验证集准确率。

- 使用Random Search来搜索最优超参数，分阶段从粗（较大超参数范围训练较少周期）到细（较小超参数范围训练较长周期）进行搜索。

**Comment by bettyeats** [comment]
https://zhuanlan.zhihu.com/p/24720954

**Comment by bettyeats** [comment]
学习率：学习率过高，容易跨过极值点，如果太小，又容易陷入局部最优。
weight_decay:不是为了提高收敛精确度，也不是提高收敛速度，而是防止过拟合。调节模型复杂度对损失函数的影响。若weight_decay很大，则复杂的模型损失函数的值也就大。
momentum:如果上一次的momentum和梯度下降的方向相同，momentum越大，梯度下降的幅度就越大。收敛的速度就会加快。
BN:将特征映射到[0,1]之间,梯度都是在1左右，避免了梯度消失,提升学习率,减少模型训练对初始化的依赖。
